amethyst_network/src/error.rs

```diff
@@ -0,0 +1,85 @@
+//! Module containing error handling logic.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    io,
+    sync::mpsc,
+};
+
+use crate::server::ServerSocketEvent;
+
+/// The `amethyst_network` result type.
+pub type Error<T> = Result<T, ErrorKind>;
+
+/// Wrapper for all errors who could occur in `amethyst_network`.
+#[derive(Debug)]
+pub enum ErrorKind {
+    /// Error that could occur on the UDP-socket
+    UdpError(laminar::error::NetworkError),
+    /// Error that could occur whit IO.
+    IoError(io::Error),
+    /// Error that could occur when serializing whit `bincode`
+    SerializeError(Box<bincode::ErrorKind>),
+    /// Error that could occur when sending an `ServerSocketEvent` to some channel.
+    ChannelSendError(mpsc::SendError<ServerSocketEvent>),
+}
+
+impl std::error::Error for ErrorKind {
+    fn description(&self) -> &str {
+        use std::error::Error;
+        match *self {
+            ErrorKind::UdpError(_) => "An UDP error occurred",
+            ErrorKind::IoError(ref error) => error.description(),
+            ErrorKind::ChannelSendError(ref error) => error.description(),
+            ErrorKind::SerializeError(ref error) => error.description(),
+        }
+    }
+
+    fn cause(&self) -> Option<&dyn std::error::Error> {
+        match *self {
+            ErrorKind::IoError(ref error) => Some(error),
+            ErrorKind::SerializeError(ref error) => Some(error),
+            ErrorKind::ChannelSendError(ref error) => Some(error),
+            _ => None,
+        }
+    }
+}
+
+impl Display for ErrorKind {
```

Comment:
I'd recommend using `err-derive` for deriving this and the `source` implementation.